### ISSUE TITLE
Deserialization to ArrayCollection not working as expected

### ISSUE THREAD

**Issue by scoolen** [schmittjoh/serializer#9]
Example:

``` php
$coll = $serializer->deserialize('["foo", "bar"]', 'ArrayCollection', 'json');
var_dump($coll);
```

Expected:

``` php
object(Doctrine\Common\Collections\ArrayCollection)[123]
  private '_elements' => 
    array
      0 => string 'foo' (length=3)
      1 => string 'bar' (length=3)
```

Got:

``` php
array
  0 => string 'foo' (length=3)
  1 => string 'bar' (length=3)
```

Moved from SerializerBundle to here: https://github.com/schmittjoh/JMSSerializerBundle/issues/255

**Comment by mlehner** [comment]
I came across this issue today while making a custom handler for a class similar to ArrayCollection.

I resolved it by cloning the visitor inside my deserialize method, which prevented the original visitor from being filled with the array result of the visitor.

Are there any known issues that would result from cloning the visitor as such?

```php
public function deserialize(VisitorInterface $visitor, $data, array $type, Context $context)
{
    $type['name'] = 'array';

    $myVisitor = clone $visitor;

    return new ScalarParameterMap($myVisitor->visitArray($data, $type, $context));
}
```

**Comment by garciall** [comment]
@schmittjoh This is happening as of 0.16.0 and @Engerim pointed at the issue already.

**Comment by AlexSayHello** [comment]
I had the same problem when implementing a handler for my collections.

I realized that the serializer was ignoring the result of my handler because the result of the visitor is not `null`
```php

// ...jms/serializer/src/JMS/Serializer/Serializer.php
private function handleDeserializeResult($visitorResult, $navigatorResult)
    {
        // This is a special case if the root is handled by a callback on the object itself.
        if (null === $visitorResult && null !== $navigatorResult) {
            return $navigatorResult;
        }

        return $visitorResult;
    }
```

What I did was re-set the navigator. This makes the result of the visitor set to `null`
```php
public function deserialize(JsonDeserializationVisitor $visitor, $data, array $type, Context $context)
    {
        /**
         * @var $collection ArrayCollection
         */
        $collection = new $type['name'];
        self::assertValidArrayCollectionSubclass($collection);

        $items = $this->getItems($visitor, $type, $data, $context);

        foreach ($items as $item) {
            $collection->add($item);
        }

        $visitor->setNavigator($visitor->getNavigator());

        return $collection;
    }
```

**Comment by goetas** [comment]
Fixed in 2.0, now the visitors do not use an internal data representation, but the data are passed always back an forward to the graph navigator and serialized/deserialized with the new SerializationVisitorInterface::getResult($data) and DeserializationVisitorInterface::getResult($data) methods
